contracts/vesting.token/src/vesting.token.cpp

```diff
@@ -0,0 +1,109 @@
+        require_auth(get_self());
+        
+        vestingtoken::vesting_settings dates;
+        dates.sales_start_date = eosio::time_point_sec::from_iso_string(sales_date_str);
+        dates.launch_date = eosio::time_point_sec::from_iso_string(launch_date_str);
+
+        vestingtoken::settings launch_sales_dates_singleton(get_self(), get_self().value);
+        launch_sales_dates_singleton.set(dates, get_self());
+    }
+
+
+    void vestingToken::assigntokens(eosio::name holder, eosio::asset amount, int category_id) {
+        // Only the contract owner can call this function
+        require_auth(get_self());
+        
+        // Check if the provided category exists in the map
+        auto category_iter = vesting_categories.find(category_id);
+        eosio::check(category_iter != vesting_categories.end(), "Invalid vesting category");
+
+        // If the category exists, you can access it like this:
+        vesting_category category = category_iter->second;
+
+
+        // Check the symbol is correct and valid
+        auto sym = amount.symbol;
+        eosio::check(sym == system_resource_currency, "Symbol does not match system resource currency");
+        eosio::check(sym.precision() == system_resource_currency.precision(), "Symbol precision does not match");
+        eosio::check(amount.amount > 0, "Amount must be greater than 0");
+
+        // Create a new vesting schedule
+        vesting_allocations vesting_table(get_self(), holder.value);
+
+        // Check the number of rows for the user
+        uint32_t num_rows = 0;
+        for (auto itr = vesting_table.begin(); itr != vesting_table.end(); ++itr) {
+            ++num_rows;
+            if (num_rows > 20) {
+                eosio::check(false, "Cannot purchase tokens more than 20 times.");
+            }
+        }
+
+        // Calculate the number of seconds since sales start
+        const eosio::time_point_sec now = eosio::current_time_point();
+        vestingtoken::settings launch_sales_dates_singleton(get_self(), get_self().value);
+        vestingtoken::vesting_settings dates = launch_sales_dates_singleton.get();
+        eosio::time_point_sec sales_start_date_value = dates.sales_start_date;
+        
+        int64_t allocated_after_sales_start_seconds = now.sec_since_epoch() - sales_start_date_value.sec_since_epoch();
+
+        vesting_table.emplace(get_self(), [&](auto& row) {
+            row.holder = holder;
+            row.total_allocated = amount;
+            row.tokens_claimed = asset(0, amount.symbol);
+            row.seconds_since_sales_start = allocated_after_sales_start_seconds;
+            row.vesting_category_type = category;
+        });
+    }
+    void vestingToken::withdraw(eosio::name holder) {
+        require_auth(holder);
+
+        // Get the vesting allocations
+        vesting_allocations vesting_table(get_self(), holder.value);
+        const eosio::time_point_sec now = eosio::current_time_point();
```

Comment:
instead do
`uint32_t now_since_epoch = eosio::time_point_sec(eosio::current_time_point()).sec_since_epoch();`
which means does not need to be calculated inside loop on L86, 90, 93. More efficient